src/renderer/dx/DxFontRenderData.cpp

```diff
@@ -0,0 +1,750 @@
+        _lineMetrics.strikethroughOffset = fullPixelAscent - _lineMetrics.strikethroughOffset;
+
+        // For double underlines we need a second offset, just below the first,
+        // but with a bit of a gap (about double the grid line width).
+        _lineMetrics.underlineOffset2 = _lineMetrics.underlineOffset +
+                                        _lineMetrics.underlineWidth +
+                                        std::round(fontSize * 0.05f);
+
+        // However, we don't want the underline to extend past the bottom of the
+        // cell, so we clamp the offset to fit just inside.
+        const auto maxUnderlineOffset = lineSpacing.height - _lineMetrics.underlineWidth;
+        _lineMetrics.underlineOffset2 = std::min(_lineMetrics.underlineOffset2, maxUnderlineOffset);
+
+        // But if the resulting gap isn't big enough even to register as a thicker
+        // line, it's better to place the second line slightly above the first.
+        if (_lineMetrics.underlineOffset2 < _lineMetrics.underlineOffset + _lineMetrics.gridlineWidth)
+        {
+            _lineMetrics.underlineOffset2 = _lineMetrics.underlineOffset - _lineMetrics.gridlineWidth;
+        }
+
+        // We also add half the stroke width to the offsets, since the line
+        // coordinates designate the center of the line.
+        _lineMetrics.underlineOffset += _lineMetrics.underlineWidth / 2.0f;
+        _lineMetrics.underlineOffset2 += _lineMetrics.underlineWidth / 2.0f;
+        _lineMetrics.strikethroughOffset += _lineMetrics.strikethroughWidth / 2.0f;
+
+        _glyphCell = actual.GetSize();
+
+        // Calculate and cache the box effect for the base font. Scale is 1.0f because the base font is exactly the scale we want already.
+        RETURN_IF_FAILED(s_CalculateBoxEffect(DefaultTextFormat().Get(), _glyphCell.width(), DefaultFontFace().Get(), 1.0f, &_boxDrawingEffect));
+    }
+    CATCH_RETURN();
+
+    return S_OK;
+}
+
+// Routine Description:
+// - Calculates the box drawing scale/translate matrix values to fit a box glyph into the cell as perfectly as possible.
+// Arguments:
+// - format - Text format used to determine line spacing (height including ascent & descent) as calculated from the base font.
+// - widthPixels - The pixel width of the available cell.
+// - face - The font face that is currently being used, may differ from the base font from the layout.
+// - fontScale -  if the given font face is going to be scaled versus the format, we need to know so we can compensate for that. pass 1.0f for no scaling.
+// - effect - Receives the effect to apply to box drawing characters. If no effect is received, special treatment isn't required.
+// Return Value:
+// - S_OK, GSL/WIL errors, DirectWrite errors, or math errors.
+[[nodiscard]] HRESULT STDMETHODCALLTYPE DxFontRenderData::s_CalculateBoxEffect(IDWriteTextFormat* format, size_t widthPixels, IDWriteFontFace1* face, float fontScale, IBoxDrawingEffect** effect) noexcept

```

Comment:
Originally `s_CalculateBoxEffect` was in `CustomTextLayout`. But I figured it's very font-related and moved it here to avoid any circle dependency.